Tasks/PythonToolInstaller/pythonToolInstaller.ts

```diff
@@ -0,0 +1,121 @@
+import * as fs from 'fs';
+import * as path from 'path';
+import * as url from 'url';
+
+import request = require('request');
+
+import * as taskLib from 'vsts-task-lib/task';
+import * as toolLib from 'vsts-task-tool-lib/tool';
+
+import { AzureStorageArtifactDownloader } from "./AzureStorageArtifacts/AzureStorageArtifactDownloader";
+import { extractCompressedFile } from 'compression-common/FileExtractor';
+
+async function run(): Promise<void> {
+    try {
+        taskLib.setResourcePath(path.join(__dirname, 'task.json'));
+        await getPython();
+    } catch (error) {
+        taskLib.error(error.message);
+        taskLib.setResult(taskLib.TaskResult.Failed, error.message);
+    }
+}
+
+async function getPython(): Promise<void> {
+    const versionSpec = taskLib.getInput('versionSpec', true);
+    const architecture = taskLib.getInput('architectureOption', true);
+    const installationSource = taskLib.getInput('installationSource', true);
+    const destination = taskLib.getPathInput('destinationDirectory', true);
+    const cleanDestination = taskLib.getBoolInput('cleanDestinationDirectory', false);
+
+    toolLib.debug('Trying to get tool from local cache first');
+    const localVersions: string[] = toolLib.findLocalToolVersions('Python');
+    const version: string = toolLib.evaluateVersions(localVersions, versionSpec);
+
+     // Clean the destination folder before downloading and extracting?
+     if (cleanDestination && fs.existsSync(destination) && fs.statSync(destination).isDirectory()) {
+        console.log(taskLib.loc('CleanDestDir', destination));
+
+        // delete the contents of the destination directory but leave the directory in place
+        for (let item of fs.readdirSync(destination)) {
+            const itemPath = path.join(destination, item);
+            taskLib.rmRF(itemPath);
+        }
+    }
+
+    if (version) {
+        // Found in tool cache. Don't set output variable.
+        console.log(taskLib.loc('Info_ResolvedToolFromCache', version));
+        return;
+    }
+
+    const compressedFile: string = await (async () => {
```

Comment:
Readability-wise, I would find a regular if-block better.  Seeing this as a function leads me to think this is going to be called outside this function's scope or passed into something that takes a function.  YMMV.

Reply:
I think that immediately-invoked functions like this are a common idiom in JS.  I grant that they leave you guessing until the end whether the function's going to get called or not, but the opening `(` before the lambda is a clue if you're familiar with the idiom.  The way I think of it, a function with no parameters is the same thing as a variable evaluated lazily.

I do like that it helps keep everything `const` though; the alternative is to `let` an undefined variable and set it in each block.  Besides the usual benefits, `const` helps out the TypeScript compiler when you switch it into strict mode (which I'll do for this file once we move the Azure downloader out).

Reply:
https://developer.mozilla.org/en-US/docs/Glossary/IIFE